File: code/typescript/src/graph.ts
import { Container, Queue, Stack } from "./containers.js";
import { NonWeightedVertex, Tag, Vertex, WeightedVertex } from "./nodes.js";

/**
 * Enumeración que define la dirección en la que se recorrerán los vértices adyacentes.
 */
export enum Direction {
  /** Recorre los vértices adyacentes en orden inverso. */
  LEFT = "LEFT",
  /** Recorre los vértices adyacentes en orden normal. */
  RIGHT = "RIGHT",
}

/**
 * Enumeración que define el algoritmo de recorrido que se utilizará para explorar el grafo.
 */
export enum Algorithm {
  /** Breadth-First Search (Búsqueda en Anchura). */
  BFS = "BFS",
  /** Depth-First Search (Búsqueda en Profundidad). */
  DFS = "DFS",
}

/**
 * Tipo que define una función que se ejecuta durante la exploración de un vértice.
 * @template T - Tipo de dato almacenado en el vértice.
 * @template Adjacency - Tipo de dato que representa la adyacencia.
 * @param vertex - El vértice que se está explorando.
 * @param arg - Argumento opcional que se puede pasar a la función.
 * @returns Un objeto que indica si la exploración debe terminar y un valor opcional.
 */
type ExploreAction<T, Adjacency> = (
  vertex: Vertex<T, Adjacency>,
  arg?: unknown,
) => { endExplore: boolean; returnValue?: unknown };

/**
 * Clase abstracta que representa un grafo genérico.
 * @template T - Tipo de dato almacenado en los vértices.
 * @template Adjacency - Tipo de dato que representa la adyacencia.
 */
export abstract class Graph<T, Adjacency> {
  /** Etiqueta o nombre del grafo. */
  protected label: string;
  /** Lista de vértices que componen el grafo. */
  protected vertexs: Vertex<T, Adjacency>[];

  /**
   * Constructor de la clase Graph.
   * @param label - Etiqueta o nombre del grafo.
   * @param vertexs - Lista opcional de vértices que componen el grafo.
   * @throws {Error} Si algún elemento en `vertexs` no es una instancia de `Vertex`.
   */
  constructor(label: string, vertexs?: Vertex<T, Adjacency>[]) {
    this.label = label;

    if (vertexs && !vertexs.every((v) => v instanceof Vertex)) {
      throw new Error("All elements in 'vertexs' must be instances of Vertex.");
    }

    this.vertexs = vertexs ? vertexs : [];
  }

  protected getRandomIntInRange(min: number, max: number): number {
    return Math.floor(Math.random() * (max - min + 1)) + min;
  }

  /**
   * Método abstracto que debe ser implementado por las clases derivadas para obtener un vértice a partir de una adyacencia.
   * @param adjacency - La adyacencia de la cual se obtendrá el vértice.
   * @returns El vértice correspondiente a la adyacencia.
   */
  protected abstract vertexFromAdjacency(
    adjacency: Adjacency,
  ): Vertex<T, Adjacency>;

  /**
   * Muestra las adyacencias de todos los vértices del grafo.
   */
  public showAdjacencies(): void {
    console.log(`Adyacencias de ${this.label}:`);

    for (const vertex of this.vertexs) {
      const adjStr = vertex.adjacencies
        .map((adj) => this.adjStr(adj))
        .join(", ");

      console.log(`${vertex} -> {${adjStr}}`);
    }

    console.log();
  }

  /**
   * Convierte una adyacencia en una cadena de texto.
   * @param adjacency - La adyacencia a convertir.
   * @returns La representación en cadena de la adyacencia.
   */
  protected adjStr(adjacency: Adjacency): string {
    return String(adjacency);
  }

  /**
   * Reinicia el estado de visitado de todos los vértices.
   */
  public resetVisited(): void {
    this.vertexs.forEach((vertex) => (vertex.visited = false));
  }

  /**
   * Imprime la información de un vértice.
   * @param vertex - El vértice a imprimir.
   * @returns Un objeto que indica que la exploración no debe terminar.
   */
  protected printAdjacency(vertex: Vertex<T, Adjacency>): {
    endExplore: boolean;
    returnValue?: unknown;
  } {
    console.log(`  ${vertex.toString()}`);
    return { endExplore: false };
  }

  /**
   * Compara dos vértices.
   * @param vertex - El vértice actual.
   * @param target - El vértice objetivo.
   * @returns Un objeto que indica si los vértices son iguales.
   * @throws {Error} Si `target` no es una instancia de `Vertex`.
   */
  protected equals(
    vertex: Vertex<T, Adjacency>,
    target?: Vertex<T, Adjacency>,
  ): {
    endExplore: boolean;
    returnValue?: unknown;
  } {
    if (!target || !(target instanceof Vertex)) {
      throw new Error("Argument should be a Vertex");
    }

    this.printAdjacency(vertex);

    return { endExplore: vertex === target };
  }

  /**
   * Explora el grafo utilizando un algoritmo específico (BFS o DFS).
   * @param start - El vértice desde el cual comenzar la exploración.
   * @param algorithm - El algoritmo a utilizar (BFS o DFS) por defecto BFS.
   * @param direction - La dirección en la que se recorrerán los vértices adyacentes por defecto RIGHT.
   * @param calcLvls - Ejecutar la función this.setLvls(start) antes de recorrer el grafo
   * @param lvlLimit - Recorrer el grafo hasta un nivel específico
   * @param iterative - Recorrer el arbol de forma iterativa, potencial comportamiento inesperado si se trata de un grafo
   * @param action - La función que se ejecutará durante la exploración por defecto undefined.
   * @param arg - Argumento opcional que se pasa a la función de acción por defecto undefined.
   * @returns El valor de retorno de la función de acción, undefined si la exploración termina fuera de la acción.
   */
  public explore({
    start,
    algorithm = Algorithm.BFS,
    direction = Direction.RIGHT,
    lvlLimit = undefined,
    calcLvls = false,
    iterative = false,
    action = undefined,
    arg = undefined,
  }: {
    start: Vertex<T, Adjacency>;
    algorithm?: Algorithm;
    lvlLimit?: number;
    calcLvls?: boolean;
    arg?: unknown;
    direction?: Direction;
    action?: ExploreAction<T, Adjacency>;
    iterative?: boolean;
  }): unknown | undefined {
    if (calcLvls) {
      this.setLvls({ root: start });
    }

    let loop = 1;
    let vertexVisited = 0;
    let vertexBeforeLoop = 0;

    const lvlLimitTitle =
      lvlLimit !== undefined ? `con límite ${lvlLimit}` : "";
    const title = `Recorriendo ${this.label} con método ${algorithm} con dirección ${direction} ${lvlLimitTitle}`;

    console.log(title);

    let vertexToCheck: Container<Vertex<T, Adjacency>> =
      algorithm === Algorithm.DFS ? new Stack() : new Queue();

    const executeAction = action || this.printAdjacency;

    vertexToCheck.add(start);
    if (!iterative) start.visited = true;
    vertexBeforeLoop += 1;

    if (iterative) console.log(`\nIteración ${loop}`);
    while (!vertexToCheck.isEmpty) {
      const currV = vertexToCheck.get();
      if (!currV) throw new Error("Unexpected null vertex");

      const { endExplore, returnValue } = executeAction(currV, arg);

      if (endExplore) {
        console.log();
        this.resetVisited();
        return returnValue;
      }

      vertexVisited += 1;

      if (iterative && vertexVisited === vertexBeforeLoop) {
        loop += 1;
        vertexToCheck = algorithm === Algorithm.DFS ? new Stack() : new Queue();
        vertexToCheck.add(start);
        vertexBeforeLoop += 1;
        vertexVisited = 0;

        console.log(`\nIteración ${loop}`);
        continue;
      }

      const adjacencies =
        direction === Direction.RIGHT
          ? currV.adjacencies
          : [...currV.adjacencies].reverse();

      for (const adjacency of adjacencies) {
        const neighbor = this.vertexFromAdjacency(adjacency as Adjacency);

        // Obviar revisión de nivel nulo en vértice
        const nLvl = neighbor.lvl as number;

        const shouldAddVertex =
          lvlLimit === undefined
            ? !neighbor.visited
            : !neighbor.visited && nLvl <= lvlLimit;

        if (shouldAddVertex) {
          vertexToCheck.add(neighbor);
          if (!iterative) neighbor.visited = true;
        }
      }
    }

    console.log();
    this.resetVisited();
    return undefined;
  }

  /**
   * Busca un vértice específico en el grafo.
   * @param start - El vértice desde el cual comenzar la búsqueda.
   * @param seek - El vértice que se está buscando.
   * @param start - El vértice desde el cual comenzar la exploración.
   * @param algorithm - El algoritmo a utilizar (BFS o DFS) por defecto BFS.
   * @param direction - La dirección en la que se recorrerán los vértices adyacentes por defecto RIGHT.
   * @param calcLvls - Ejecutar la función this.setLvls(start) antes de recorrer el grafo
   * @param lvlLimit - Recorrer el grafo hasta un nivel específico
   * @param iterative - Recorrer el arbol de forma iterativa, potencial comportamiento inesperado si se trata de un grafo
   * @returns número de nodos visitados antes de encontrarlo, undefined si no se encontró el nodo buscado
   */
  public seek({
    start,
    seek,
    algorithm = Algorithm.BFS,
    direction = Direction.RIGHT,
    lvlLimit = undefined,
    iterative = false,
    calcLvls = false,
    eq = (v1, v2) => v1 === v2,
  }: {
    start: Vertex<T, Adjacency>;
    seek: Vertex<T, Adjacency>;
    algorithm?: Algorithm;
    direction?: Direction;
    calcLvls?: boolean;
    lvlLimit?: number;
    iterative?: boolean;
    eq?: (v1: Vertex<T, Adjacency>, v2: Vertex<T, Adjacency>) => boolean;
  }): number | undefined {
    const nodesVisited = [0];
    const search = this.explore({
      start,
      algorithm,
      arg: nodesVisited,
      direction,
      lvlLimit,
      iterative,
      calcLvls,
      action: (vertex, arg) => {
        const nodesVisited = arg as Array<number>;
        nodesVisited[0] += 1;

        if (!seek || !(seek instanceof Vertex)) {
          throw new Error("Argument should be a Vertex");
        }

        this.printAdjacency(vertex);

        return { endExplore: eq(start, seek), returnValue: nodesVisited[0] };
      },
    });

    return search as number | undefined;
  }

  /**
   * Calcula los niveles de los vértices a partir de un vértice raíz.
   * @param root - El vértice raíz desde el cual calcular los niveles.
   * @param direction - La dirección en la que se recorrerán los vértices adyacentes por defecto RIGHT.
   */
  public setLvls({
    root,
    direction = Direction.RIGHT,
  }: {
    root: Vertex<T, Adjacency>;
    direction?: Direction;
  }) {
    const vertexToCheck = new Queue<Vertex<T, Adjacency>>();
    console.log(`Calculando niveles...`);
    vertexToCheck.add(root);
    root.visited = true;
    root.lvl = 1;

    while (!vertexToCheck.isEmpty) {
      const currV = vertexToCheck.get();
      if (!currV || !currV.lvl) throw new Error("Unexpected behavior");

      console.log(`  ${currV.toString()}`);
      const adjacencies =
        direction === Direction.RIGHT
          ? currV.adjacencies
          : [...currV.adjacencies].reverse();

      for (const adj of adjacencies) {
        const vertex = this.vertexFromAdjacency(adj);
        if (!vertex.visited) {
          vertex.visited = true;
          vertex.lvl = currV.lvl + 1;
          vertexToCheck.add(vertex);
        }
      }
    }

    console.log();
    this.resetVisited();
  }

  public exploreWithHeuristic({
    start,
    seek,
    arg,
    heuristic,
    action = (vertex, seek) => {
      console.log(vertex.toString());
      return { endExplore: vertex == seek };
    },
  }: {
    start: Vertex<T, Adjacency>;
    seek: Vertex<T, Adjacency>;
    arg?: { [key: string]: unknown };
    action?: (
      currV: Vertex<T, Adjacency>,
      seek: Vertex<T, Adjacency>,
      arg?: { [key: string]: unknown },
    ) => { endExplore: boolean; returnValue?: unknown };
    heuristic: (
      adj: Vertex<T, Adjacency>,
      currV: Vertex<T, Adjacency>,
      seek: Vertex<T, Adjacency>,
      arg?: { [key: string]: unknown },
    ) => number;
  }) {
    let agenda: Array<Vertex<T, Adjacency>> = [];
    agenda.push(start);

    while (agenda.length != 0) {
      const currV = agenda.pop();

      if (currV == undefined) {
        throw new Error("fisrtElement should not be undefined");
      }

      const { endExplore, returnValue } = action(currV, seek, arg);

      if (endExplore) {
        return returnValue;
      }

      for (const adjacency of currV.adjacencies) {
        const adj = this.vertexFromAdjacency(adjacency);
        agenda.push(adj);
      }

      if (agenda.length != 0) {
        // Ordenamos la agenda usando la heurística
        agenda.sort(
          (a, b) =>
            heuristic(a, currV, seek, arg) - heuristic(b, currV, seek, arg),
        );

        // Obtenemos el valor mínimo de la heurística
        const minHeuristic = heuristic(agenda[0], currV, seek, arg);

        // Filtramos los nodos que tienen la heurística mínima
        const minSolutions = agenda.filter(
          (v) => heuristic(v, currV, seek, arg) === minHeuristic,
        );

        const choosenIndex = this.getRandomIntInRange(
          0,
          minSolutions.length - 1,
        );

        const choosenOpt = minSolutions[choosenIndex];
        agenda = [choosenOpt];
      } else {
        agenda = [start];
      }
    }
  }
}

/**
 * Clase que representa un grafo no ponderado.
 * @template T - Tipo de dato almacenado en los vértices.
 */
export class NonWeightedGraph<T> extends Graph<T, NonWeightedVertex<T>> {
  /**
   * Devuelve el vértice directamente, ya que no hay pesos involucrados.
   * @param adjacency - La adyacencia de la cual se obtendrá el vértice.
   * @returns El vértice correspondiente a la adyacencia.
   */
  protected vertexFromAdjacency(
    adjacency: NonWeightedVertex<T>,
  ): NonWeightedVertex<T> {
    return adjacency;
  }
}

/**
 * Clase que representa un grafo ponderado.
 * @template T - Tipo de dato almacenado en los vértices.
 */
export class WeightedGraph<T> extends Graph<T, [WeightedVertex<T>, number]> {
  /**
   * Devuelve el vértice de una adyacencia, ignorando el peso.
   * @param adjacency - La adyacencia de la cual se obtendrá el vértice.
   * @returns El vértice correspondiente a la adyacencia.
   */
  protected vertexFromAdjacency(
    adjacency: [WeightedVertex<T>, number],
  ): WeightedVertex<T> {
    const [vertex] = adjacency;
    return vertex;
  }

  /**
   * Calcula el camino más corto desde un vértice dejando una etiqueta en cada vertice
   * del grafo el nodo a seguir para llegar al destino.
   * @param destiny - El vértice destino desde el cual calcular el camino más corto.
   * @param direction - La dirección en la que se recorrerán los vértices adyacentes.
   */
  public shortestPath({
    destiny,
    direction = Direction.RIGHT,
  }: {
    destiny: WeightedVertex<T>;
    direction?: Direction;
  }) {
    const vertexToCheck = new Queue<WeightedVertex<T>>();
    console.log(`Etiquetando...`);
    vertexToCheck.add(destiny);
    destiny.visited = true;
    destiny.tag = {
      vertex: undefined,
      weigth: 0,
    };

    while (!vertexToCheck.isEmpty) {
      const currV = vertexToCheck.get();
      if (!currV || !currV.lvl) throw new Error("Unexpected behavior");

      console.log(`  ${currV.toString()}`);
      const adjacencies =
        direction === Direction.RIGHT
          ? currV.adjacencies
          : [...currV.adjacencies].reverse();

      for (const adj of adjacencies) {
        const [vertex, weigth] = adj;
        const tag = { vertex, weigth };

        if (!vertex.visited) {
          vertex.visited = true;
          vertex.tag = tag;
          vertexToCheck.add(vertex);
        }

        const prevTag = vertex.tag as Tag<T>;
        if (vertex.tag !== undefined && tag.weigth < prevTag.weigth) {
          vertex.tag = tag;
        }
      }
    }

    this.resetVisited();
  }

  /**
   * Convierte una adyacencia en una cadena de texto que incluye el peso.
   * @param adjacency - La adyacencia a convertir.
   * @returns La representación en cadena de la adyacencia.
   */
  protected adjStr(adjacency: [WeightedVertex<T>, number]): string {
    const [vertex, weight] = adjacency;
    return `(${vertex.value}, ${weight})`;
  }
}

File: code/python/con_tipado/graph.py
from abc import ABC, abstractmethod
from enum import Enum, auto
from random import randint
from typing import Any, Callable, Generic, Optional, TypeVar

from containers import Container, Queue, Stack
from nodes import NonWeightedVertex, Vertex, WeightedVertex

T = TypeVar("T")
Adjacency = TypeVar("Adjacency")


class Graph(Generic[T, Adjacency], ABC):
    """
    Clase abstracta que representa un grafo genérico con tipos parametrizados.

    Atributos:
        label (str): Etiqueta descriptiva del grafo
        vertexs (list[Vertex[T, Adjacency]]): Lista de vértices del grafo

    Args:
        T: Tipo de dato almacenado en los vértices
        Adjacency: Tipo de las relaciones de adyacencia

    Raises:
        ValueError: Si se proporcionan vértices que no son instancias de Vertex
    """

    class Direction(Enum):
        """Enumera las direcciones de recorrido posibles"""

        LEFT = auto()  # Recorrido por la izquierda (adyacencias invertidas)
        RIGHT = auto()  # Recorrido por la derecha (orden natural)

    class Algorithm(Enum):
        """Enumera los algoritmos de recorrido disponibles"""

        BFS = auto()  # Breadth-First Search
        DFS = auto()  # Depth-First Search

    class Objective(Enum):
        MINIMIZE = auto()
        MAXIMIZE = auto()

    def __init__(
        self, label: str, vertexs: Optional[list[Vertex[T, Adjacency]]] = None
    ) -> None:
        """
        Inicializa un nuevo grafo.

        Args:
            label: Nombre identificativo del grafo
            vertexs: Lista opcional de vértices iniciales
        """
        self.label = label

        if vertexs and not all(isinstance(v, Vertex) for v in vertexs):
            raise ValueError("All elements in 'vertexs' must be instances of Vertex.")

        self.vertexs: list[Vertex[T, Adjacency]] = vertexs if vertexs else []

    @abstractmethod
    def vertex_from_adjacency(self, adjacency: Adjacency) -> Vertex[T, Adjacency]:
        """
        Método abstracto para obtener el vértice destino desde una adyacencia.

        Args:
            adjacency: Relación de adyacencia a convertir

        Returns:
            Vértice correspondiente a la adyacencia
        """
        pass

    def adj_str(self, adjacency: Adjacency) -> str:
        """
        Formatea una adyacencia para representación como cadena.

        Args:
            adjacency: Relación de adyacencia a formatear

        Returns:
            Representación en cadena de la adyacencia
        """
        return str(adjacency)

    def show_adjacencies(self) -> None:
        """Muestra todas las adyacencias del grafo en formato legible."""
        print(f"Adyacencias de {self.label}:")
        for vertex in self.vertexs:
            adj_str = ", ".join(self.adj_str(adj) for adj in vertex.adjacencies)
            print(f"{vertex} -> {{{adj_str}}}")
        print()

    def reset_visited(self) -> None:
        """Reinicia el estado de visitado de todos los vértices."""
        for vertex in self.vertexs:
            vertex.visited = False

    def print_adjacency(
        self,
        vertex: Vertex[T, Adjacency],
        _: Optional[Any],
    ) -> tuple[bool, Optional[Any]]:
        """
        Acción por defecto para imprimir vértices durante el recorrido.

        Args:
            vertex: Vértice actual siendo visitado
            _: Argumento adicional no utilizado

        Returns:
            Tupla (False, None) para continuar el recorrido
        """
        print(f"  {vertex}")
        return (False, None)

    def equals(
        self, vertex: Vertex, vertex2: Optional[Any]
    ) -> tuple[bool, Optional[Any]]:
        """
        Compara dos vértices para búsqueda.

        Args:
            vertex: Vértice actual siendo visitado
            vertex2: Vértice buscado como objetivo

        Returns:
            Tupla (True, None) si coinciden, (False, None) en caso contrario

        Raises:
            ValueError: Si vertex2 no es una instancia de Vertex
        """
        if vertex2 is None or not isinstance(vertex2, Vertex):
            raise ValueError("Arg should be a Vertex")

        self.print_adjacency(vertex, None)

        return (True, None) if vertex is vertex2 else (False, None)

    def explore(
        self,
        start: Vertex[T, Adjacency],
        algorithm: Algorithm,
        direction: Direction = Direction.RIGHT,
        lvl_limit: Optional[int] = None,
        set_lvls: bool = False,
        iterative: bool = False,
        action: Optional[
            Callable[[Vertex[T, Adjacency], Optional[Any]], tuple[bool, Optional[Any]]]
        ] = None,
        arg: Optional[Any] = None,
    ) -> Optional[Any]:
        """
        Realiza un recorrido parametrizado del grafo ejecutando lógica personalizada en cada vértice.

        Parámetros:
            start: Vértice raíz donde inicia el recorrido
            algorithm: Estrategia de recorrido (Algoritmo.BFS o Algoritmo.DFS)
            direction: Orden de procesamiento de adyacencias (LEFT=invertido, RIGHT=natural)
            lvl_limit: Limitar la busqueda a un nivel específico, para esto los vértices deben
                tener el nivel puesto de forma correcta, si no se estableció ningún nivel la función
                soltará una excepción durante el recorrido o si se establecen niveles mal puede ocurrir
                un comportamiento inesperado
            set_lvls: Ejecutar la función self.set_lvls(start) antes de empezar la exploración
            iterative: Recorrer un arbol de forma iterativa (potencial error de ciclado en grafos)
            action: Función callback con firma:
                   (vértice_actual, arg) -> (detener_recorrido: bool, valor_retorno: Any)
                   - Si detener_recorrido = True, se aborta el recorrido y retorna valor_retorno
                   - Si detener_recorrido = False, continúa normalmente
            arg: Argumento opcional que se pasa a la función 'action'

        Retorno:
            - Valor retornado por 'action' si detiene el recorrido
            - None si completa todo el recorrido sin interrupciones
        """
        if set_lvls:
            self.set_lvls(start)

        loop = 1
        vertex_visited = 0
        vertex_before_loop = 0

        vertex_to_check: Container[Vertex[T, Adjacency]] = (
            Stack() if algorithm == self.Algorithm.DFS else Queue()
        )

        if action is None:
            action = self.print_adjacency
        assert action is not None

        limitTitle = f"con límite {lvl_limit}" if lvl_limit is not None else ""
        print(f"Recorrido {algorithm.name} por {direction.name} {limitTitle}")

        vertex_to_check.add(start)
        vertex_before_loop += 1
        if not iterative:
            start.visited = True

        while not vertex_to_check.is_empty():
            curr_v = vertex_to_check.get()

            if curr_v is None:
                raise RuntimeError("Vertice actual es None")

            end_explore, value_return = action(curr_v, arg)

            if end_explore:
                print()
                self.reset_visited()
                return value_return

            vertex_visited += 1

            if iterative and vertex_visited == vertex_before_loop:
                loop += 1
                vertex_to_check = (
                    Stack() if algorithm == self.Algorithm.DFS else Queue()
                )
                vertex_to_check.add(start)
                vertex_before_loop += 1
                vertex_visited = 0
                print(f"\nIteración {loop}")
                continue

            adjacencies = (
                curr_v.adjacencies
                if direction == self.Direction.RIGHT
                else list(reversed(curr_v.adjacencies))
            )

            for adjacency in adjacencies:
                neighbor = self.vertex_from_adjacency(adjacency)

                if lvl_limit is not None:
                    assert neighbor.lvl
                    should_add = not neighbor.visited and neighbor.lvl <= lvl_limit

                else:
                    should_add = not neighbor.visited

                if should_add:
                    vertex_to_check.add(neighbor)
                    if not iterative:
                        neighbor.visited = True

        print()
        self.reset_visited()
        return None

    def seek(
        self,
        start: Vertex[T, Adjacency],
        seek: Vertex[T, Adjacency],
        algorithm: Algorithm = Algorithm.BFS,
        direction: Direction = Direction.RIGHT,
        lvl_limit: Optional[int] = None,
        set_lvls: bool = False,
        iterative: bool = False,
        eval_eq: Callable[
            [Vertex[T, Adjacency], Vertex[T, Adjacency]], bool
        ] = lambda v1, v2: v1
        == v2,
    ) -> Optional[int]:
        """
        Busca un vértice en el grafo mediante recorrido.

        Args:
            start: Vértice inicial de búsqueda
            seek: Vértice objetivo a encontrar
            algorithm: Algoritmo de búsqueda a utilizar
            direction: Dirección de procesamiento de adyacencias
            direction: Orden de procesamiento de adyacencias (LEFT=invertido, RIGHT=natural)
            lvl_limit: Limitar la busqueda a un nivel específico, para esto los vértices deben
                tener el nivel puesto de forma correcta, si no se estableció ningún nivel la función
                soltará una excepción durante el recorrido o si se establecen niveles mal puede ocurrir
                un comportamiento inesperado
            set_lvls: Ejecutar la función self.set_lvls(start) antes de empezar la exploración

        Returns:
            None cuando encuentra el vértice o si no existe
        """
        nodesVisited = [0]

        def action(v: Vertex[T, Adjacency], arg) -> tuple[bool, Any]:
            arg[0] += 1

            print(f"{v}")
            return (eval_eq(v, seek), arg[0])

        print(f"Buscando {seek}")
        return self.explore(
            start=start,
            algorithm=algorithm,
            direction=direction,
            action=action,
            arg=nodesVisited,
            lvl_limit=lvl_limit,
            set_lvls=set_lvls,
            iterative=iterative,
        )

    def set_lvls(
        self, root: Vertex[T, Adjacency], direction: Direction = Direction.RIGHT
    ):
        vertex_to_check = Queue[Vertex[T, Adjacency]]()
        print("Calculando niveles...")
        root.visited = True
        root.lvl = 1
        vertex_to_check.enqueue(root)

        while not vertex_to_check.is_empty():
            curr_v = vertex_to_check.get()
            assert curr_v is not None
            assert curr_v.lvl is not None

            print(f"  {curr_v}")

            adjacencies = (
                curr_v.adjacencies
                if direction == self.Direction.RIGHT
                else list(reversed(curr_v.adjacencies))
            )

            for adj in adjacencies:
                vertex = self.vertex_from_adjacency(adj)
                if not vertex.visited:
                    vertex.visited = True
                    vertex.lvl = curr_v.lvl + 1
                    vertex_to_check.add(vertex)

        print()
        self.reset_visited()

    def hill_climbing(
        self,
        start: Vertex[T, Adjacency],
        seek: Vertex[T, Adjacency],
        action: Callable[
            [
                Vertex[T, Adjacency],  # curr_v
                Vertex[T, Adjacency],  # seek
                Any,  # arg
            ],
            tuple[bool, Optional[Any]],  # (end_explore, return)
        ],
        heuristic: Callable[
            [
                Vertex[T, Adjacency],  # adjacency
                Vertex[T, Adjacency],  # curr_v
                Vertex[T, Adjacency],  # seek
                Any,  # arg
            ],
            float,
        ],
        objective: Objective = Objective.MINIMIZE,
        arg: Optional[Any] = None,
    ):
        print("Recorrido Hill Climbing")
        reverse = False if objective == self.Objective.MINIMIZE else True
        agenda: list[Vertex[T, Adjacency]] = []
        agenda.append(start)
        while not len(agenda) == 0:
            curr_v = agenda.pop()
            assert curr_v is not None

            end_explore, returnValue = action(curr_v, seek, arg)

            if end_explore:
                return returnValue

            for adjacency in curr_v.adjacencies:
                vertex = self.vertex_from_adjacency(adjacency)
                agenda.append(vertex)

            if len(agenda) != 0:
                # Ordenar la agenda usando la heurística
                agenda.sort(
                    key=lambda v: heuristic(v, curr_v, seek, arg), reverse=reverse
                )

                # Obtener el valor mínimo de la heurística
                choosen_heuristic = heuristic(agenda[0], curr_v, seek, arg)

                # Filtrar las soluciones con la heurística mínima
                min_solutions = [
                    v
                    for v in agenda
                    if heuristic(v, curr_v, seek, arg) == choosen_heuristic
                ]

                # Elegir una solución
                choosen_index = randint(0, len(min_solutions) - 1)
                choosen_opt = min_solutions[choosen_index]
                agenda = [choosen_opt]
            else:
                print("No se llegó a la solución")


class NonWeightedGraph(Graph[T, NonWeightedVertex[T]]):
    """
    Grafo no ponderado donde las adyacencias son vértices directamente.

    Hereda de:
        Graph[T, NonWeightedVertex[T]]
    """

    def vertex_from_adjacency(
        self, adjacency: NonWeightedVertex[T]
    ) -> NonWeightedVertex[T]:
        """
        Obtiene el vértice destino desde una adyacencia no ponderada.

        Args:
            adjacency: Vértice adyacente directo

        Returns:
            El mismo vértice de la adyacencia
        """
        return adjacency


class WeightedGraph(Graph[T, tuple[WeightedVertex[T], float]]):
    """
    Grafo ponderado donde las adyacencias son tuplas (vértice, peso).

    Hereda de:
        Graph[T, tuple[WeightedVertex[T], float]]
    """

    def vertex_from_adjacency(
        self, adjacency: tuple[WeightedVertex[T], float]
    ) -> WeightedVertex[T]:
        """
        Obtiene el vértice destino desde una adyacencia ponderada.

        Args:
            adjacency: Tupla (vértice adyacente, peso)

        Returns:
            Componente vértice de la tupla de adyacencia
        """
        vertex, _ = adjacency
        return vertex

    def adj_str(self, adjacency: tuple[WeightedVertex[T], float]) -> str:
        """
        Formatea una adyacencia ponderada como '(valor, peso)'.

        Args:
            adjacency: Tupla (vértice, peso) a formatear

        Returns:
            Cadena formateada con valor del vértice y peso
        """
        vertex, weight = adjacency
        return f"({vertex.value}, {weight})"
